doc/developer/design/20230721_persist_fast_path.md

```diff
@@ -0,0 +1,81 @@
+
+## Context
+
+When is a query like `SELECT * FROM relation LIMIT n` fast?
+
+In the general case, a query spins up an anonymous dataflow that writes its results into an arrangement. Once the arrangement has caught up to a specific frontier, we read the data back out of the arrangement, apply the limit, and return it the user.  If `relation` is an arbitrary view, this could be arbitrarily expensive.
+
+However, if `relation` is indexed, this query behaves very differently. There’s no need to spin up a dataflow or build a new arrangement: we can just read the data from the pre-existing arrangement that’s maintained by the index. This type of “fast path” query can only apply very simple transformations on the indexed data: maps, filters, projections, limits, and a couple others. Anything more complex falls back to the more general dataflow-based query machinery.
+
+Users often seem to expect this same sort of behaviour when `relation` is a durable collection, like a source or a materialized view. This is understandable: in an ordinary database, this query might just need to scan through the first few entries in a file before returning a result. However, these sorts of queries currently fall back to the general-case logic, spinning up a full dataflow that reads in the entire collection just to filter it down to a few rows. This design proposes adding a new fast path, similar to the index-based fast-path, that can make these sorts of queries more performant.
+
+## Goals
+
+- Fast responses to `SELECT * FROM large_collection LIMIT small_n` queries.
+
+## Non-Goals
+
+- Improving the performance of `LIMIT` clause on a more general range of queries.
+
+## Overview
+
+Supporting a fast query path for select-limit queries against Persist shards will require changes to both our query planning and execution.
+
+- In `environmentd`, generalize the existing fast-path planning logic. As of today, fast-path plans were only generated when the queried collection has at least one index. We’ll extend this logic to also generate these plans for unindexed collections that are backed by a Persist shard.
+- When `clusterd` receives a peek against a Persist-backed collection — as opposed to an arrangement, which is always the case today — it will create a new `PersistPeek` operation that streams the data from a Persist shard. While it’s possible to efficiently and incrementally stream consolidated data from a Persist shard, no such API exists today; we’ll need to add one.
+
+## Detailed description
+
+### Planning
+
+Today, `environmentd` will generate a fast-path plan when a dataflow can be represented as: a `Get` from an indexed collection, a map-filter-project stage, and some “finishing” logic including order-by/limit. We’ll change this to also allow the `Get` to get from a non-indexed collection backed by a Persist shard… as long as the MFP stage does not include any filters.
```

Comment:
I am wondering here if we can, for a start, be even more restrictive than only banning filters.

For example, wouldn't queries such as `SELECT attribute1 FROM persist_collection_with_many_attributes` still potentially do unbounded work as described below? Would such queries also be parallelized across workers as is the case today with `persist_source` if processed on the fast path? 

Similar questions might apply to `ORDER BY`, since in general you could be ordering by different attributes than what the persist shard is organized by?

Reply:
> For example, wouldn't queries such as `SELECT attribute1 FROM persist_collection_with_many_attributes` still potentially do unbounded work as described below?

A couple perspectives on that query:
- It does not contain a limit clause, so the potential result set size is unbounded. If you wrote a query like this against an index today, we would eventually hit a limit in terms of the maximum number of bytes returned. It seems to me like a similar backstop for a Persist-backed query would also make sense.
- It selects just one column out of supposedly-many columns, so the work done per byte returned is larger. (Though the work done per row returned is the same.) This is a constant factor for any particular table or query, and I haven't given these much thought. It may make sense to have a bytes-read bound for a particular query (perhaps controlled by some system var) which could let us bound this much more explicitly.

> Similar questions might apply to ORDER BY, since in general you could be ordering by different attributes than what the persist shard is organized by?

Yes, that makes sense to me as well - nice catch.

Reply:
> > For example, wouldn't queries such as `SELECT attribute1 FROM persist_collection_with_many_attributes` still potentially do unbounded work as described below?
> 
> A couple perspectives on that query:
> 
> * It does not contain a limit clause, so the potential result set size is unbounded. If you wrote a query like this against an index today, we would eventually hit a limit in terms of the maximum number of bytes returned. It seems to me like a similar backstop for a Persist-backed query would also make sense.

It's a good point that eventually we will hit the ceiling on maximum number of bytes returned. Another guardrail here is the statement timeout, so it is especially important -- as you already hinted at by using asynchrony with a task-per-peek approach -- to be able to cancel processing effectively. It is also attractive to do so so as to not tie up workers, since they need to keep IVM/peek work going. With peeks from indexes, in particular, these concerns are somewhat mitigated by: (a) data is in memory already, so there are no stalls to read from storage; (b) we can use multi-worker parallelism for peeks.

> It selects just one column out of supposedly-many columns, so the work done per byte returned is larger. (Though the work done per row returned is the same.) This is a constant factor for any particular table or query, and I haven't given these much thought. It may make sense to have a bytes-read bound for a particular query (perhaps controlled by some system var) which could let us bound this much more explicitly.

Perhaps a crazy idea: If Persist is collecting statistics during compaction, one could envision computing a conservative lower bound on the number of rows that would be returned by a full scan (e.g., by just knowing how many additions / retractions exist across batches). That might allow you to know that some queries cannot be processed at all without spending any effort.

Reply:
Not crazy! We have update-count stats already, and have proposed a net row count stat in https://github.com/MaterializeInc/materialize/issues/20618 - seems totally plausible that we could eventually leverage such things.

Though the earlier we decide we don't want to send a query down the fast path, the better... it would be regrettable if a query that previously would have succeeded gets sent down the fast path, hits some limit, and errors.